federation/pkg/federation-controller/replicaset/planner/planer.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package planer
+
+import (
+	"sort"
+
+	fed_api "k8s.io/kubernetes/federation/apis/federation"
+)
+
+// Planer decides how many out of the given replicas should be placed in each of the
+// federated clusters.
+type Planer struct {
+	preferences *fed_api.FederatedReplicaSetPreferences
+}
+
+type namedLocalReplicaSetPreferences struct {
+	clusterName string
+	fed_api.LocalReplicaSetPreferences
+}
+
+type byWeight []*namedLocalReplicaSetPreferences
+
+func (a byWeight) Len() int      { return len(a) }
+func (a byWeight) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+
+// Preferences are sorted according by decreasing weight and increasing clusterName.
```

Comment:
We have to sort by something to ensure that the plan is stable. I picked weight+name because it helps to decide who to give more replicas if the global replica count doesn't allow to fulfill all minReplica requests. Imagine that one cluster has minReplicas to 2 and the other minReplicas 1. Which one is in more need of replicas?